File: cpp_solutions/Tree/Path_Sum_II.cpp
/*
* Given a binary tree and a sum, find all root-to-leaf paths 
* where each path's sum equals the given sum.
*
**/

 bool Sum(TreeNode* node, int sum, int target, vector<int> temp,
             vector<vector<int>>& rst){
	if(node == NULL)
		return false;
	temp.push_back(node->val);
	sum = sum + node->val;
	if(node->left == NULL && node->right == NULL)
	{
		if(sum == target){
            rst.push_back(temp);
			return true;
        }
		else
			return false;
	}

	bool left = Sum(node->left, sum, target, temp, rst);
	bool right = Sum(node->right, sum, target, temp, rst);

	return left || right;
}


vector<vector<int>> pathSum(TreeNode* root, int sum) {
      
    vector<vector<int>> rst;
    vector<int> temp;  
    if(root == NULL)
      return rst;
      
    if(Sum(root, 0, sum, temp, rst))
       return rst;
     
    return rst;
}

File: python_solutions/Tree/Path_Sum_II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        self.ret = []
        self.solution = []
        
        self.findPath(self.ret, self.solution, root, sum)
        return self.ret
    
    def findPath(self, ret, solution, root, sum):
        if root is None:
            return
        sum = sum - root.val
        if root.left is None and root.right is None:
            if sum == 0:
                solution.append(root.val)
                ret.append(list(solution))
                solution.pop(-1)
            return
        solution.append(root.val)
        self.findPath(ret, solution, root.left, sum)
        self.findPath(ret, solution, root.right, sum)
        solution.pop(-1)
                
        